File: python/1400.py
# Link: https://leetcode.com/problems/construct-k-palindrome-strings/description/
# Runtime: 20 ms
# Memory: 18.15 MB


class Solution:
    def canConstruct(self, s: str, k: int) -> bool:
        if len(s) < k:
            return False
        elif len(s) == k:
            return True

        pairs = Counter(s).most_common()
        print
        number_of_odds = 0
        for pair in pairs:
            if pair[1] % 2:
                number_of_odds += 1

        return number_of_odds <= k

File: c++/1400.cpp
// Link: https://leetcode.com/problems/construct-k-palindrome-strings/description/
// Runtime: 4 ms
// Memory: 13.54 MB

class Solution {
public:
  bool canConstruct(const std::string &s, int k) {
    if (static_cast<int>(s.size()) < k)
      return false;
    else if (static_cast<int>(s.size()) == k)
      return true;

    std::vector<int> frequencies(26);
    for (const auto &symbol : s)
      frequencies[symbol - 'a']++;

    int numberOfOdds = 0;
    for (const auto &val : frequencies)
      if (val % 2)
        numberOfOdds++;

    return numberOfOdds <= k;
  }
};
